File: backend/src/routes/dict-options.ts
import { Router, Request, Response } from 'express';
import { PrismaClient } from '@prisma/client';

const prisma = new PrismaClient();
const router = Router();

// GET /api/dict-options?categoryId= — Get dropdown options by category
router.get('/', async (req: Request, res: Response) => {
  try {
    const { categoryId, lang } = req.query;
    const where: any = {};
    if (categoryId) where.category_id = parseInt(categoryId as string);
    if (lang)       where.lang = lang as string;
    else            where.lang = 'en'; // default to English

    const data = await prisma.dict_option.findMany({
      where,
      orderBy: { option_id: 'asc' },
    });

    res.json(data.map(d => ({
      categoryId:  d.category_id,
      optionId:    d.option_id,
      option:      d.option,
      lang:        d.lang,
      description: d.description,
      status:      d.status,
    })));
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: 'Failed to fetch dict options' });
  }
});

// GET /api/dict-options/categories — Get all distinct categories
router.get('/categories', async (_req: Request, res: Response) => {
  try {
    const data = await prisma.dict_option.findMany({
      where: { lang: 'en' },
      distinct: ['category_id'],
      select: { category_id: true, description: true },
      orderBy: { category_id: 'asc' },
    });

    res.json(data.map(d => ({
      categoryId: d.category_id,
      description: d.description,
    })));
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: 'Failed to fetch categories' });
  }
});

export { router as dictOptionRoutes };

File: backend/app/routers/dict_options.py
"""Dictionary option endpoints."""
from fastapi import APIRouter, Depends, Query
from fastapi.responses import JSONResponse
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import text

from app.database import get_db

router = APIRouter()


@router.get("")
async def get_dict_options(
    categoryId: int | None = Query(None),
    lang: str | None = Query(None),
    db: AsyncSession = Depends(get_db),
):
    try:
        conditions = []
        params: dict = {}

        if categoryId is not None:
            conditions.append("category_id = :category_id")
            params["category_id"] = categoryId

        effective_lang = lang if lang else "en"
        conditions.append("lang = :lang")
        params["lang"] = effective_lang

        where_clause = " AND ".join(conditions) if conditions else "1=1"

        result = await db.execute(
            text(f"SELECT * FROM eam.dict_option WHERE {where_clause} ORDER BY option_id ASC"),
            params,
        )
        rows = result.mappings().all()

        return [
            {
                "categoryId": r["category_id"],
                "optionId": r["option_id"],
                "option": r["option"],
                "lang": r["lang"],
                "description": r["description"],
                "status": r["status"],
            }
            for r in rows
        ]
    except Exception as e:
        print(f"Error: {e}")
        return JSONResponse(status_code=500, content={"error": "Failed to fetch dict options"})


@router.get("/categories")
async def get_categories(db: AsyncSession = Depends(get_db)):
    try:
        result = await db.execute(
            text("""
                SELECT DISTINCT ON (category_id) category_id, description
                FROM eam.dict_option
                WHERE lang = 'en'
                ORDER BY category_id ASC
            """)
        )
        rows = result.mappings().all()

        return [
            {
                "categoryId": r["category_id"],
                "description": r["description"],
            }
            for r in rows
        ]
    except Exception as e:
        print(f"Error: {e}")
        return JSONResponse(status_code=500, content={"error": "Failed to fetch categories"})
